File: P_2/Examen_P2/Napakalaki_Java_P2/src/napakalaki/BadConsequence.java
// MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA

package napakalaki;
import java.util.ArrayList; // Importamos el util ArrayList para poder usarlo

/**
 *
 * @author ixjosemi
 */

public class BadConsequence
{
    // Declaramos los atributos privados a utilizar
    static int MAXTREASURES = 10;
    private String text; // Representa lo que dice un mal rollo
    private int levels; // Representa el numero de niveles que se pierden
    private int nVisibleTreasures; // Nº Tesoros visibles que se pierden
    private int nHiddenTreasures; // Nº Tesoros ocultos que se pierden
    private boolean death; // Representa un mal rollo de tipo muerte


    // Constructor que inicializa a valores por defecto
    public BadConsequence(String text, int levels, int nVisible, int nHidden)
    {
        this.text = text;
        this.levels = levels;
        this.nVisibleTreasures = nVisible;
        this.nHiddenTreasures = nHidden;
    }

    // Constructor que inicializa a valores por defecto
    public BadConsequence (String text, boolean death)
    {
        this.text = text;
        this.death = death;
    }

    // Listas de HiddenTreasures y VisibleTreasures
    private ArrayList <TreasureKind> specificHiddenTreasures = new ArrayList();

    private ArrayList <TreasureKind> specificVisibleTreasures = new ArrayList();

    // Constructor que inicializa a valores por defecto
    public BadConsequence(String text, int levels, ArrayList<TreasureKind> tVisible,
            ArrayList<TreasureKind> tHidden)
    {
        this.text = text;
        this.levels = levels;
        this.specificHiddenTreasures = tVisible;
        this.specificVisibleTreasures = tHidden;
    }

    // Consultores básicos
    // Devuelve text
    public String getText()
    {
        return text;
    }

    // Devuelve levels
    public int getLevels()
    {
        return levels;
    }

    // Devuelve VisibleTreasures
    public int getVisibleTreasures()
    {
        return nVisibleTreasures;
    }

    // Devuelve HiddenTreasures
    public int getHiddenTreasures()
    {
        return nHiddenTreasures;
    }

    // Devuelve si estás muerto o no
    public boolean getDeath()
    {
        return death;
    }

    // Devuelve el SpecificHIddenTreasure
    public ArrayList<TreasureKind> getSpecificHiddenTreasures()
    {
        return specificHiddenTreasures;
    }

    // Devuelve el SpecificVisibleTreasure
    public ArrayList<TreasureKind> getSpecificVisibleTreasures()
    {
        return specificVisibleTreasures;
    }

    public boolean isEmpty(){
        boolean empty = true;

        if(this.levels == 0 && this.death == false && this.nHiddenTreasures == 0
                && this.nVisibleTreasures == 0 && this.specificHiddenTreasures.isEmpty()
                && this.specificVisibleTreasures.isEmpty())
            empty = true;

        return empty;
    }

    public void substractVisibleTreasures(Treasure t){

    }

    public void substractHiddenTreasures(Treasure t){

    }
    /*
    public BadConsequence adjustToFitTreasureLists(Treasure v, Treasure h){

    }
    */

    // Método toString para pintar una secuencia
    public String toString()
    {
        return  text + " \nLost levels = " + levels
            + " \nVisible treasures = " + nVisibleTreasures
            + " \nHidden treasures = " + nHiddenTreasures
            + " \nDeath? = " + death
            + " \nSpecific Hidden Treasures = " + specificHiddenTreasures
            + " \nSpecific Visible Treasures = " + specificVisibleTreasures;
    }
}

File: P_3/examen_P3/Napakalaki_Ruby_P3/lib/bad_consequence.rb
#encoding: utf-8
# MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA


module NapakalakiRuby
    
class BadConsequence

    # Consultores básicos
    @@MAXTREASURES = 10
    
    attr_reader :aText
    attr_reader :someLevels
    private_class_method :new

    # Metodo initalize que inicia los atributos
    def initialize(aText = "", someLevels = 0, nVisibleTreasures = 0,
                nHiddenTreasures = 0,  someSpecificVisibleTreasures = Array.new,
                someSpecificHiddenTreasures = Array.new, death = false)
        @aText = aText
        @someLevels = someLevels
        @nVisibleTreasures = nVisibleTreasures
        @nHiddenTreasures = nHiddenTreasures
        @someSpecificVisibleTreasures = someSpecificVisibleTreasures
        @someSpecificHiddenTreasures = someSpecificHiddenTreasures
        @death = death
    end

    # Constructor #1
    def self.newLevelNumberOfTreasures(aText, someLevels,
                      nVisibleTreasures, nHiddenTreasures)
        new(aText, someLevels, nVisibleTreasures, nHiddenTreasures)
    end

  # Constructor #2
    def self.newLevelSpecificTreasures(aText, someLevels,
                      someSpecificVisibleTreasures, someSpecificHiddenTreasures)
    new(aText, someLevels, 0, 0, someSpecificVisibleTreasures, someSpecificHiddenTreasures,
        false)
    end

    # Constructor #3
    def self.newDeath (aText)
        new(aText, 0, 0, 0, Array.new, Array.new, true)
    end

    def is_empty
        empty = false
        
        if (@levels == 0 && @death == false && @nHiddenTreasures == 0 &&
           @nVisibleTreasures == 0 && @someSpecificVisibleTreasures.empty? &&
           @someSpecificHiddenTreasures.empty?)
                empty = true
        end
        return empty
    end
    
    def substract_visible_treasure(treasure) 
        @specificVisibleTreasures.delete(treasure.type);
    end
     
    def substract_hidden_treasure(treasure)
        @specificHiddenTreasures.delete(treasure.type);      
    end
    
    def adjust_to_fit_treasure_list(v, h)
       
        t_visible = Array.new
        t_hidden = Array.new
         
        v.each do |t|
     
            if (t_visible.index(t.type) == nil)
                t_visible << t.type
            end
        end
         
        h.each do |t|

            if (t_hidden.index(t.type) == nil)
                t_hidden << t.type
            end
        end
        bc = BadConsequence.newLevelSpecificTreasures(@text, 0, t_visible, t_hidden)
        
        return bc  
    end
    
    # Metodo to_string que pinta en pantalla lo deseado
    def to_s
        "#{@aText}
        Niveles perdidos: #{@someLevels.to_s}
        Objetos visibles perdidos: #{@nVisibleTreasures.to_s}
        Objetos ocultos perdidos: #{@nHiddenTreasures.to_s}
        Objetos especificos visibles: #{@someSpecificVisibleTreasures}
        Objetos especificos ocultos: #{@someSpecificHiddenTreasures}
        Muerto? #{@death.to_s} \n"
    end
 
end
end
